src/Infrastructure/Process/FileCopier/SymfonyFileCopier.php

```diff
@@ -0,0 +1,63 @@
+<?php
+
+namespace PhpTuf\ComposerStager\Infrastructure\Process\FileCopier;
+
+use PhpTuf\ComposerStager\Domain\Output\ProcessOutputCallbackInterface;
+use PhpTuf\ComposerStager\Exception\DirectoryNotFoundException;
+use PhpTuf\ComposerStager\Exception\ProcessFailedException;
+use RecursiveCallbackFilterIterator;
+use RecursiveDirectoryIterator;
+use SplFileInfo;
+use Symfony\Component\Filesystem\Exception\IOException;
+use Symfony\Component\Filesystem\Filesystem;
+use UnexpectedValueException;
+
+/**
+ * @internal
+ */
+final class SymfonyFileCopier implements SymfonyFileCopierInterface
+{
+    /**
+     * @var \Symfony\Component\Filesystem\Filesystem
+     */
+    private $filesystem;
+
+    public function __construct(Filesystem $filesystem)
+    {
+        $this->filesystem = $filesystem;
+    }
+
+    public function copy(string $from, string $to, array $exclusions = [], ?ProcessOutputCallbackInterface $callback = null): void
+    {
+        try {
+            $iterator = $this->createIterator($from);
+            $this->filesystem->mirror($from, $to, $iterator);
+        } catch (IOException $e) {
+            throw new ProcessFailedException($e->getMessage(), (int) $e->getCode(), $e);
+        }
+    }
+
+    /**
+     * @throws \PhpTuf\ComposerStager\Exception\DirectoryNotFoundException
+     */
+    private function createIterator(string $from): RecursiveCallbackFilterIterator
+    {
+        try {
+            $directoryIterator = new RecursiveDirectoryIterator($from);
+        } catch (UnexpectedValueException $e) {
+            throw new DirectoryNotFoundException($from, 'The "copy from" directory does not exist at "%s"');
```

Comment:
Passing the previous exception as well can aid with debugging.

Reply:
Thanks, I keep missing forgetting to do that. I'll see if I can write a static analysis check for it for future.